src/bio/tests/wal_ut.c

```diff
@@ -0,0 +1,815 @@
+{
+	int	rc, ret;
+
+	uuid_generate(args->bua_pool_id);
+	rc = bio_mc_create(args->bua_xs_ctxt, args->bua_pool_id, meta_sz, wal_sz, data_sz, 0);
+	if (rc) {
+		D_ERROR("UT MC create failed. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+	/* Sleep 1 second to ensure each test uses unique WAL re-format timestamps */
+	sleep(1);
+
+	rc = bio_mc_open(args->bua_xs_ctxt, args->bua_pool_id, 0, &args->bua_mc);
+	if (rc) {
+		D_ERROR("UT MC open failed. "DF_RC"\n", DP_RC(rc));
+		ret = bio_mc_destroy(args->bua_xs_ctxt, args->bua_pool_id, 0);
+		if (ret)
+			D_ERROR("UT MC destroy failed. "DF_RC"\n", DP_RC(ret));
+	}
+
+	return rc;
+}
+
+struct ut_fake_tx {
+	uint32_t		ft_act_max;
+	uint32_t		ft_buf_sz;
+	uint32_t		ft_act_nr;
+	uint32_t		ft_payload_sz;
+	uint32_t		ft_act_idx;
+	uint32_t		ft_copy_ptr_sz;
+	struct umem_action	*ft_acts;
+	char			*ft_buffer;
+	unsigned int		ft_ckp_done:1;
+};
+
+static uint32_t
+fake_act_nr(struct umem_wal_tx *tx)
+{
+	struct ut_fake_tx	*fake_tx = (struct ut_fake_tx *)&tx->utx_private;
+
+	return fake_tx->ft_act_nr;
+}
+
+static uint32_t
+fake_payload_sz(struct umem_wal_tx *tx)
+{
+	struct ut_fake_tx	*fake_tx = (struct ut_fake_tx *)&tx->utx_private;
+
+	return fake_tx->ft_payload_sz;
+}
+
+static struct umem_action *
+fake_act_first(struct umem_wal_tx *tx)
+{
+	struct ut_fake_tx	*fake_tx = (struct ut_fake_tx *)&tx->utx_private;
+
+	fake_tx->ft_act_idx = 0;
+	return &fake_tx->ft_acts[fake_tx->ft_act_idx];
+}
+
+static struct umem_action *
+fake_act_next(struct umem_wal_tx *tx)
+{
+	struct ut_fake_tx	*fake_tx = (struct ut_fake_tx *)&tx->utx_private;
+
+	fake_tx->ft_act_idx++;
+	if (fake_tx->ft_act_idx >= fake_tx->ft_act_nr)
+		return NULL;
+	return &fake_tx->ft_acts[fake_tx->ft_act_idx];
+}
+
+static struct umem_wal_tx_ops ut_fake_wal_tx_ops = {
+	.wtx_act_nr	= fake_act_nr,
+	.wtx_payload_sz	= fake_payload_sz,
+	.wtx_act_first	= fake_act_first,
+	.wtx_act_next	= fake_act_next,
+};
+
+static void
+ut_tx_free(struct umem_wal_tx *fake_wal_tx)
+{
+	struct ut_fake_tx	*fake_tx;
+
+	fake_tx = (struct ut_fake_tx *)&fake_wal_tx->utx_private;
+	D_FREE(fake_tx->ft_acts);
+	D_FREE(fake_tx->ft_buffer);
+
+	D_FREE(fake_wal_tx);
+}
+
+static struct umem_wal_tx *
+ut_tx_alloc(uint32_t act_nr, uint32_t buf_sz)
+{
+	struct umem_wal_tx	*fake_wal_tx;
+	struct ut_fake_tx	*fake_tx;
+
+	D_ALLOC_PTR(fake_wal_tx);
+	assert_non_null(fake_wal_tx);
+	fake_wal_tx->utx_ops = &ut_fake_wal_tx_ops;
+
+	fake_tx = (struct ut_fake_tx *)&fake_wal_tx->utx_private;
+
+	D_ALLOC_ARRAY(fake_tx->ft_acts, act_nr);
+	assert_non_null(fake_tx->ft_acts);
+	fake_tx->ft_act_max = act_nr;
+
+	if (buf_sz) {
+		D_ALLOC(fake_tx->ft_buffer, buf_sz);
+		assert_non_null(fake_tx->ft_buffer);
+		fake_tx->ft_buf_sz = buf_sz;
+		dts_buf_render(fake_tx->ft_buffer, buf_sz);
+	}
+
+	return fake_wal_tx;
+}
+
+static inline uint64_t
+rand_addr(void)
+{
+	uint64_t addr  = rand();
+
+	return (addr << 32) + rand();
```

Comment:
maybe just addr << 32 + addr is fine.

Reply:
well, use two different rand number looks better to me.